File: server/app/services/llm/embedding.py
from typing import List
from openai import OpenAI
from app.models.embedding_config import EmbeddingConfig
from app.logger import logger


class EmbeddingService:
    @staticmethod
    def create_client(embedding_config: EmbeddingConfig) -> OpenAI:
        return OpenAI(
            api_key=embedding_config.api_key,
            base_url=embedding_config.base_url
        )

    @staticmethod
    def embed_texts_sync(
        embedding_config: EmbeddingConfig,
        texts: List[str]
    ) -> List[List[float]]:
        client = EmbeddingService.create_client(embedding_config)
        try:
            response = client.embeddings.create(
                model=embedding_config.model_id,
                input=texts
            )
            vectors = [item.embedding for item in response.data]
            logger.info(f"Embedded {len(texts)} texts successfully")
            return vectors
        except Exception as e:
            logger.error(f"Failed to embed texts: {str(e)}", exc_info=True)
            raise

    @staticmethod
    def embed_query_sync(
        embedding_config: EmbeddingConfig,
        query: str
    ) -> List[float]:
        client = EmbeddingService.create_client(embedding_config)
        try:
            response = client.embeddings.create(
                model=embedding_config.model_id,
                input=query
            )
            vector = response.data[0].embedding
            logger.info(f"Embedded query successfully")
            return vector
        except Exception as e:
            logger.error(f"Failed to embed query: {str(e)}", exc_info=True)
            raise

    @staticmethod
    async def embed_texts(
        embedding_config: EmbeddingConfig,
        texts: List[str]
    ) -> List[List[float]]:
        from openai import AsyncOpenAI
        client = AsyncOpenAI(
            api_key=embedding_config.api_key,
            base_url=embedding_config.base_url
        )
        try:
            response = await client.embeddings.create(
                model=embedding_config.model_id,
                input=texts
            )
            vectors = [item.embedding for item in response.data]
            logger.info(f"Embedded {len(texts)} texts successfully")
            return vectors
        except Exception as e:
            logger.error(f"Failed to embed texts: {str(e)}", exc_info=True)
            raise

    @staticmethod
    async def embed_query(
        embedding_config: EmbeddingConfig,
        query: str
    ) -> List[float]:
        from openai import AsyncOpenAI
        client = AsyncOpenAI(
            api_key=embedding_config.api_key,
            base_url=embedding_config.base_url
        )
        try:
            response = await client.embeddings.create(
                model=embedding_config.model_id,
                input=query
            )
            vector = response.data[0].embedding
            logger.info(f"Embedded query successfully")
            return vector
        except Exception as e:
            logger.error(f"Failed to embed query: {str(e)}", exc_info=True)
            raise

File: server_go/internal/service/llm/embedding.go
package llm

import (
	"context"
	"fmt"

	openai "github.com/sashabaranov/go-openai"
)

type EmbeddingService struct {
	client  *openai.Client
	modelID string
	dim     int
}

func NewEmbeddingService(baseURL, apiKey, modelID string, dim int) *EmbeddingService {
	cfg := openai.DefaultConfig(apiKey)
	if baseURL != "" {
		cfg.BaseURL = baseURL
	}
	return &EmbeddingService{
		client:  openai.NewClientWithConfig(cfg),
		modelID: modelID,
		dim:     dim,
	}
}

func (es *EmbeddingService) Embed(ctx context.Context, texts []string) ([][]float32, error) {
	resp, err := es.client.CreateEmbeddings(ctx, openai.EmbeddingRequestStrings{
		Input: texts,
		Model: openai.EmbeddingModel(es.modelID),
	})
	if err != nil {
		return nil, fmt.Errorf("embedding request failed: %w", err)
	}

	result := make([][]float32, 0, len(resp.Data))
	for _, d := range resp.Data {
		result = append(result, d.Embedding)
	}

	return result, nil
}

func (es *EmbeddingService) EmbedSingle(ctx context.Context, text string) ([]float32, error) {
	embeddings, err := es.Embed(ctx, []string{text})
	if err != nil {
		return nil, err
	}
	if len(embeddings) == 0 {
		return nil, fmt.Errorf("no embedding returned")
	}
	return embeddings[0], nil
}

func (es *EmbeddingService) Dimension() int {
	return es.dim
}
